### ISSUE TITLE
import/extend functionality in TD

### ISSUE THREAD

**Issue by sebastiankb** [w3c/wot-thing-description#168]
As original initiated by @Kazuo-Kajimoto and reactivated again in Bundang we should bring in an import / extend functionality in the TD:

Lets assume there is a master TD template providing all basic interactions and a huge set of (identical) Things implements these interactions. It would be beneficial to maintain only the master TD and all the Things would inherit the interactions. Thing’s TD only extends its individual information (e.g., id, base URI, semantics like locations, etc) or overwrite master TD information. Example:

![image](https://user-images.githubusercontent.com/13832739/42951321-3b1805b6-8b76-11e8-8436-b8e5dafea6fd.png)

Open questions:

- Security aspect: Can we trust a master TD?
- Who construct the "full" TD? Lets assume when a client request a TD. Does it get the full resolved TD or the client has to download the master TD separately to get the full TD?

**Comment by vcharpenay** [comment]
There are two analogies I think of around this idea:
 - when a Node.js module imports another one, the corresponding source file is dynamically loaded in memory and evaluated, so that the program can transparently use it.
 - HTML does not have an explicit import tag. Instead, it is often the case that a template engine renders the full document before it is exposed to the client.

Neither requires a standard import mechanism (in Node.js, import is implemented in JavaScript, it is not part of the interpreter). I would therefore argue that having TDs extending a master TD should be up to WoT implementers. Only fully interpretable TDs should be exchanged between servients. The exchange overhead that comes along is probably not significant and even if it is, it could be solved at the protocol or serialization level (as with Protocol Buffers or schema-informed EXI).

However, for maintenance purposes, a TD could refer by a link to the master TD (or 'template') that was used to generate it, like:
```js
{
  "links": [
    { "href": "...", "rel": "td:extends" }
  ]
}
```
and the relation type `td:extends` could be part of the TD model. There also exists several well-defined terms that one may reuse for that purpose, like [`prov:wasDerivedFrom`](https://www.w3.org/TR/2013/REC-prov-o-20130430/#wasDerivedFrom) or [`dc:source`](http://dublincore.org/documents/dcmi-terms/#terms-source).

**Comment by benfrancis** [comment]
@sebastiankb wrote:
> Who construct the "full" TD? Lets assume when a client request a TD. Does it get the full resolved TD or the client has to download the master TD separately to get the full TD?

Given the complexity this would add to a client if it had to individually fetch multiple resources and resolve the inheritance tree in order to arrive at a useful Thing Description, it would make more sense for the server to do all of this internally and expose a complete Thing Description to the client.

@vcharpenay wrote:
>  I would therefore argue that having TDs extending a master TD should be up to WoT implementers. Only fully interpretable TDs should be exchanged between servients.

I agree. This need not be part of the TD specification, it can just be an implementation detail of a Web of Things server which could use a simple templating mechanism or OOP inheritance.

**Comment by mkovatsc** [comment]
To my understanding, this mechanism was originally proposed for managing Thing information as part of the product lifecycle and management, i.e., for "WoT implementers" as @vcharpenay says, but not to use it to share TDs with WoT Clients.

We should clarify again the understanding and requirements among the Participants.

I share the opinion of @benfrancis and @vcharpenay that this should not be used for constructing TDs on the client side. Hence, we should leave it out of scope.

A templating mechanism for product management and alignment within vertical consortia (e.g., for home appliances) as in the [original proposal](https://github.com/w3c/wot/tree/master/proposals/td-lifecycle) appears to be a worthy item for the next charter period. It helps Thing manufacturers and fosters convergence.

**Comment by ToruKawaguchi** [comment]
I fully agree @mkovatsc , @benfrancis and @vcharpenay that this should be not something that WoT Clients need to handle.

> To my understanding, this mechanism was originally proposed for managing Thing information as part of the product lifecycle and management, i.e., for "WoT implementers" as @vcharpenay says, but not to use it to share TDs with WoT Clients.

I checked @Kazuo-Kajimoto about his intention and he confirmed that above text exactly reflects his thought.

**Comment by takuki** [comment]
It would be useful if Thing Directory has a capability to store TD templates that themselves are still invalid as TD, and allow to create an offspring new TD from those templates to make a complete TD. We can start experimenting ideas such as this in WoT IG first, and I personally think we can make it into WoT TD V2 in the next charter period.

**Comment by sebastiankb** [comment]
Good to see that there is a common agreement about the self-contained TD. 

I like @takuki idea. Another idea can be to create a "TD Management Note" in the future that addresses the topics like TD lifecycle, Thing Directory, etc. and can be used as a guideline for implementer.

**Comment by mlagally** [comment]
There's one scenario, where a "Client" would use these TD templates and fill them with life: 
Before a device is actually built, it is useful to create simulations based on a template. 
In the case of the Oracle Digital Twin simulator we follow this approach, so it can be seen 
as a kind of "universal client".

My initial thought is to keep the protocol binding (forms) out of the template, 
i.e. it just describes the abstract interface, i.e. interaction endpoints and events (i.e. message formats).

A regular device client only needs to expose an (instance-) TD which contains values and 
concrete protocol bindings.

**Comment by mkovatsc** [comment]
I would argue that this is still a server use case: it is the simulator that takes the abstract TD and instantiates it with simulated Interaction resources.

And yes, forms would be omitted, as they carry concrete metadata of an instance. This is also how the Scripting API does it.

**Comment by sebastiankb** [comment]
this is definitely something that is not covered in this TD version. Maybe a topic for the next version.

**Comment by sebastiankb** [comment]
based on today's TD call, @kolotu will provide an example how Eclipse Vorto has implement the **extend** functionality

**Comment by kolotu** [comment]
The extend functionality is implemented for [Functionblocks](https://github.com/eclipse/vorto/blob/development/docs/vortolang-1.0.md#function-block) in Eclipse Vorto. Functionblocks are a set of properties and capabilities for a device and would be close to an equivalent of Thing Description Templates in WoT. 

Functionblocks can extend other Functionblocks explicitly or they can be extended, when they are "instantiated" in an [InformationModel](https://github.com/eclipse/vorto/blob/development/docs/vortolang-1.0.md#information-model). 

In the example below the InfomationModel [IM](https://vorto-dev.eclipse.org/#/details/com.kevin.official.test.examples:IM:1.0.1) uses the Functionblock FbC and extends it inline. The Functionblock [FbC](https://vorto-dev.eclipse.org/#/details/com.kevin.official.test.examples:FbC:1.0.1) extends [FbB](https://vorto-dev.eclipse.org/#/details/com.kevin.official.test.examples:FbB:1.0.1). FbB extends [FbA](https://vorto-dev.eclipse.org/#/details/com.kevin.official.test.examples:FbA:1.0.0).

![inheritanceExample](https://user-images.githubusercontent.com/56577255/80370354-3e4b1e80-8890-11ea-9ab5-951298cbee33.png)

Overriding properties is not allowed, so it wouldn't be possible to introduce another property a in FbB. This also prevents naming collisions.

**Comment by sebastiankb** [comment]
In the PR #1024 I like to introduce the "extend" feature for TM for the next draft.
 
@kolotu if I understand correctly overwriting is not possible in Vorto. That means FbB cannot override the "a" with type string with another type, right? What is the motivation for not doing so?

**Comment by kolotu** [comment]
@sebastiankb I think the reasoning behind not allowing overriding of properties was to preserve the semantics throughout all hierarchies. If the top-level defines that all instances should have a certain property with a certain type, all APIs should be able to rely on that.

**Comment by egekorkan** [comment]
I do have a problem on understanding what is the real use case behind defining how TMs are generated. We do not define a way how TMs are managed/stored/accessed etc. so defining how they are generated (maybe from other TMs) seem a bit strange to me. My opinion would be that we should first have a well understood way to generate TDs from TMs and then explore downwards from that point on.

**Comment by egekorkan** [comment]
Another point, this can be difficult to impossible without constraining what a TM could be. How can a TM extend another one if it does not have the required protocol binding or simply the IP address? At the moment, TMs are allowed to have form elements to any degree of precision.

**Comment by JKRhb** [comment]
> Another point, this can be difficult to impossible without constraining what a TM could be. How can a TM extend another one if it does not have the required protocol binding or simply the IP address? At the moment, TMs are allowed to have form elements to any degree of precision.

Maybe this could be addressed (in part) by not allowing for a `base` and restricting `hrefs` in `forms` to relative URIs?

(I hope it is okay if comment on this topic, I am currently working with @citrullin on an implementation of WoT TD as a module for RIOT OS and we are quite interested in the potential the Thing Model concept has to offer in this context.)

**Comment by JKRhb** [comment]
I just noticed that this issue could probably be closed, considering that #1024 has landed by now.

**Comment by sebastiankb** [comment]
since we introduced the Thing Model concept in TD 1.1 which also supports import / extend functionality we can close issue
